source/sema.h

```diff
@@ -19,23 +19,147 @@
 #define CPP2_SEMA_H
 
 #include "reflect.h"
+#include "dll.h"
 
 
 namespace cpp2 {
 
+static reflect_libraries* active_instance = nullptr;
+
+class reflect_libraries
+{
+    std::vector<error_entry>& errors;
+    std::string               in_progress = {};
+    std::vector<dll>          loaded      = {};
+
+    struct function_entry
+    {
+        std::variant<
+            std::monostate,
+            meta::mf_sign_type_in*,
+            meta::mf_sign_type_inout*
+        > type = {};
+        std::variant<
+            std::monostate,
+            meta::mf_sign_func_in*,
+            meta::mf_sign_func_inout*
+        > func = {};
```

Comment:
This approach is very type safe, but requires all the other methods to distinguish between function and and type metafunctions. Would it be possible to have an enum for type, function, etc. We could then define a tuple that could be accessed via the enum value. 

There could even be a consexpr lookup from the argument type  of `try_apply` to the index in the tuple.

This would remove a lot of code duplication and would make it simpler to add new metafunction for different kinds, e.g. using decl, variable decl, member function decl, member variable decl, namespace dcl, etc.

Reply:
Yes, we need a better approach what would avoid a lot of code duplication for what is essentially "overload handling". I'll keep it in mind when designing the full name look-up solution.

Reply:
👍